core-bundle/src/Util/PackageUtil.php

```diff
@@ -14,13 +14,20 @@
 
 use PackageVersions\Versions;
 
+/**
+ * @deprecated Deprecated since Contao 4.13, to be removed in Contao 5.0; use
+ *             the Composer\InstalledVersions class instead
+ */
 class PackageUtil
 {
     /**
      * Returns the version number of a package.
      */
     public static function getVersion(string $packageName): string
     {
+        trigger_deprecation('contao/core-bundle', '4.13', 'Using the PackageUtil::getVersion() method has been deprecated and will no longer work in Contao 5.0. Use the Composer\InstalledVersions class instead.');
```

Comment:
shouldn't the deprecation be triggered in the class itself and not in every method?

Reply:
These are all static methods…

Reply:
Yes, but the deprecation is still triggered once the autoloader loads the class, is it not?

Reply:
I don't want that. I want the deprecations to be triggered if a method is used, not if the class is autoloaded. Also, the deprecations have different messages, so…

Reply:
> I don't want that. I want the deprecations to be triggered if a method is used, not if the class is autoloaded.

Why not, we have already used that strategy for a lot of classes. The deprecation should be triggered as soon as anyone uses the class, regardless of which method.

Reply:
but that's what we usually do if the whole class is deprecated …
https://github.com/contao/contao/blob/27f248764c86bd9929ec1b961c9241b5e593c9e2/core-bundle/src/EventListener/UserAwareTrait.php
https://github.com/contao/contao/blob/89833ae383bda4d86d60d909b01c3fe7d13edbb9/core-bundle/src/Resources/contao/controllers/FrontendCron.php
https://github.com/contao/contao/blob/ef97b65f6965e78c8cf6a824dbcd0dc28e9e4f72/installation-bundle/src/Database/AbstractVersionUpdate.php
https://github.com/contao/contao/blob/d66886594592f41a11084286e20bebf314a1c1f6/core-bundle/src/Resources/contao/elements/ContentMarkdown.php
https://github.com/contao/contao/blob/b8ce2592bd15750c0adbffcf5d829c7c107328b3/core-bundle/src/Framework/ScopeAwareTrait.php

Reply:
Jeez… Why are we even discussing this? We are just waisting time.

Some of the deprecations say "use Composer\InstalledVersions" and others say "use ContaoCoreBundle::getVersion()". How am I supposed to distinguish between the two at autoloading time?